src/entities/trade.rs

```diff
@@ -549,5 +549,685 @@ mod tests {
             assert_eq!(trade.input_amount.currency, TOKEN0.clone());
             assert_eq!(trade.output_amount.currency, ETHER.clone());
         }
+
+        #[test]
+        fn it_can_be_constructed_with_ether_as_input_for_exact_output() {
+            let trade = Trade::new(
+                Route::new(vec![PAIR_WETH_0.clone()], ETHER.clone(), TOKEN0.clone()),
+                CurrencyAmount::from_raw_amount(TOKEN0.clone(), 100).unwrap(),
+                TradeType::ExactOutput,
+            )
+            .unwrap();
+            assert_eq!(trade.input_amount.currency, ETHER.clone());
+            assert_eq!(trade.output_amount.currency, TOKEN0.clone());
+        }
+
+        #[test]
+        fn it_can_be_constructed_with_ether_as_output_for_exact_input() {
+            let trade = Trade::new(
+                Route::new(vec![PAIR_WETH_0.clone()], TOKEN0.clone(), ETHER.clone()),
+                CurrencyAmount::from_raw_amount(TOKEN0.clone(), 100).unwrap(),
+                TradeType::ExactInput,
+            )
+            .unwrap();
+            assert_eq!(trade.input_amount.currency, TOKEN0.clone());
+            assert_eq!(trade.output_amount.currency, ETHER.clone());
+        }
+
+        #[test]
+        #[should_panic(expected = "PAIRS")]
+        fn best_trade_exact_in_throws_with_empty_pairs() {
```

Comment:
```suggestion
    }

    mod best_trade_exact_in {
        use super::*;

        #[test]
        #[should_panic(expected = "PAIRS")]
        fn throws_with_empty_pairs() {
```
Why don't you follow what's in v2-sdk but instead put these tests in `mod new`?